juju/framework.py

```diff
@@ -320,6 +328,8 @@ def notices(self, event_path):
 
 class Framework:
 
+    on = FrameworkEvents()
```

Comment:
The framework has no handle. How is this working at all?

We need proper and careful testing for all of that new logic.

Reply:
I suppose because for `BoundEvent`s `self.emitter` is a subclass of `EventsBase`.

For example: `<BoundEvent StartEvent bound to CharmEvents.start at 0x7f056779b128>`

So when `emit` is called on a `BoundEvent`, the `emitter` has a handle because it is a subclass of `EventsBase` which is an `Object` with a handle:

```
        event = self.event_type(Handle(self.emitter, self.event_kind, key), *args, **kwargs)
        framework._emit(event)
```

Unfortunately, `emitter` has different meaning in different places currently:

* in the `__get__` function for `EventsBase` the `emitter` is the `charm Object` if `charm_obj.on` was accessed (it depends on whether you access a class or an object properly);
* in the `emit()` call of `BoundEvent` the emitter is a subclass of `EventsBase`. `EventsBase` has `key=None` in its `__init__` so unless you specify one, the same `event_type` will always have fixed parent handle (such as "MyCharm/on").

So, in the `Framework` case the emitter in `emit` of `BoundEvent` is `FrameworkEvents(EventsBase)`.

Reply:
So, as a result, the handle for `pre_commit` will look like `on/pre_commit` because `self.emitter` in `BoundEvent` is not `Framework` - it is `FrameworkEvents` instead.

```
(Pdb++) l 154
149 B	    def emit(self, *args, **kwargs):
150  	        """Emit event to all registered observers.
151  	
152  	        The current storage state is committed before and after each observer is notified.
153  	        """
154  ->	        framework = self.emitter.framework
155  	        # TODO This needs to be persisted.
156  	        framework._event_count += 1
157  	        key = str(framework._event_count)
158  	        event = self.event_type(Handle(self.emitter, self.event_kind, key), *args, **kwargs)
159  	        framework._emit(event)
(Pdb++) self.emitter
<juju.framework.FrameworkEvents object at 0x7fd3127a7490>

# ...

(Pdb++) Handle(self.emitter, self.event_kind, key)
<juju.framework.Handle object at 0x7fd31264b610>
(Pdb++) print(Handle(self.emitter, self.event_kind, key))
on/pre_commit[7]
```

Reply:
As pointed out on the channel, all of that is referring to the "on" events value. We know what it is.. the issue is that it needs to be rooted on something well defined, and right now it's not. The parent there will be None, which means we have the "on" of None.. doesn't make sense. This is a framework event, and the handle should reflect that.

Reply:
Updated the PR with an additional commit to address this.